Add required built-ins for PSyclone 2.0.0 (for #1107)

TeranIvy opened the pull request:
As described in #1107, this PR adds new built-ins gathered from the LFRic requirements for PSyclone 2.0.0.

Built-ins that take `real`-valued fields as arguments:
- `a_plus_X(Y,a,X)` where `Y = a + X`; `Y`, `X` are fields and `a` is a scalar (all `real`-valued).
-  `inc_a_plus_X(a,X)` where `X = a + X`; `X` is a field and `a` is a scalar (all `real`-valued).
- `aX_plus_aY(Z,a,X,Y)` where `Z = a*X + a*Y = a*(X + Y)`; `Z`, `Y`, `X` are fields and `a` is a scalar (all `real`-valued).
- `aX_minus_bY(Z,a,X,b,Y)` where `Z = aX - bY`; `Z`, `Y`, `X` are fields and `a`, `b` are scalars (all `real`-valued).
-  `a_divideby_X(Y,a,X)` where `Y = a/X`; `Y`, `X` are fields and `a` is a scalar (all `real`-valued).
-  `inc_a_divideby_X(a,X)` where `X = a/X`; `X` is a field and `a` is a scalar (all `real`-valued).
- `sign_X(Y,a,X)` where `Y = sign(a,X)`; `Y`, `X` are fields and `a` is a scalar (all `real`-valued). *Note: Carried over from #560.* 
- `real2int_X(Y, X)` where `Y = int(X, i_def)`;  `X` is a `real`-valued field and `Y` is an `integer`-valued field.

Built-ins that take `integer`-valued fields as arguments: 
- `int_a_plus_X(Y,a,X)` where `Y = a + X`; `Y`, `X` are fields and `a` is a scalar (all `integer`-valued).
- `int_inc_a_plus_X(a,X)` where `X = a + X`; `X` is a field and `a` is a scalar (all `integer`-valued).
- `int_sign_X(Y,a,X)` where `Y = sign(a,X)`; `Y`, `X` are fields and `a` is a scalar (all `integer`-valued).
- `int2real_X(Y, X)` where `Y = real(X, r_def)`; `X` is an `integer`-valued field and `Y` is a `real`-valued field.

This PR also:
- Adds a check that the mixing of `integer`-valued and `real`-valued fields is only allowed for the conversion built-ins `real2int_X(Y, X)` and `int2real_X(Y, X)`;
- Tidies up inconsistencies in code generation.

arporter commented on src/psyclone/tests/test_files/dynamo0p3/15.28.1_int_sign_X_builtin.f90:
```diff
@@ -0,0 +1,52 @@
+  implicit none
+
+  type(integer_field_type) :: f1, f2
+  real(i_def)              :: a
```
s/real/integer/ I think?

TeranIvy replied:
Oops, thanks for spotting! Done.

arporter commented on src/psyclone/tests/test_files/dynamo0p3/invalid_builtins_mod.f90:
```diff
@@ -143,8 +143,19 @@ module lfric_builtins_mod
 
-  ! TODO in #1107: Add a check that mixing will only be allowed for the
-  ! built-ins that convert 'real'- to 'integer'-valued fields and vice-versa.
+  !> Invalid built-in that has different data type of its field
```
"...that has field arguments of different primitive types and is not a data-type conversion..."

TeranIvy replied:
That reads better, thanks. Changed.

arporter reviewed (changes requested):
All tests and examples OK with compilation.
Updated docs build fine.
I've checked all files for pycodestyle/pylint.
Nothing major - the documentation just needs re-wording/correcting in a few places.

TeranIvy commented on src/psyclone/tests/domain/lfric/lfric_builtins_test.py:
```diff
@@ -274,6 +274,37 @@ def test_builtin_args_not_same_space():
+
+    '''
+    # Save the name of the actual built-in-definitions file
+    old_name = lfric_builtins.BUILTIN_DEFINITIONS_FILE[:]
```
I learned something new, thanks 😄 Changed.

arporter replied:
I only learned this recently :-)

TeranIvy commented on src/psyclone/tests/domain/lfric/lfric_builtins_test.py:
```diff
@@ -483,6 +514,142 @@ def test_inc_X_plus_Y(tmpdir, monkeypatch, annexed, dist_mem):
+
+    if not dist_mem:
+        output = (
+            "    SUBROUTINE invoke_0(a, f1)\n"
```
Agreed, done.

TeranIvy commented on src/psyclone/tests/domain/lfric/lfric_builtins_test.py:
```diff
@@ -483,6 +514,142 @@ def test_inc_X_plus_Y(tmpdir, monkeypatch, annexed, dist_mem):
+
+    if not dist_mem:
+        output = (
+            "    SUBROUTINE invoke_0(f2, a, f1)\n"
```
Yep, done.

TeranIvy commented on src/psyclone/tests/domain/lfric/lfric_builtins_test.py:
```diff
@@ -837,6 +1004,77 @@ def test_inc_aX_plus_bY(tmpdir, monkeypatch, annexed, dist_mem):
+
+    if not dist_mem:
+        output = (
+            "    SUBROUTINE invoke_0(f3, a, f1, f2)\n"
```
Done.

TeranIvy commented on src/psyclone/tests/domain/lfric/lfric_builtins_test.py:
```diff
@@ -1618,6 +1927,142 @@ def test_inc_X_divideby_Y(tmpdir, monkeypatch, annexed, dist_mem):
+    if not dist_mem:
+        output = (
+            "    SUBROUTINE invoke_0(a, f1, b, f2, f3)\n"
+            "      REAL(KIND=r_def), intent(in) :: a, b\n"
```
Done.

TeranIvy commented on src/psyclone/tests/domain/lfric/lfric_builtins_test.py:
```diff
@@ -2058,15 +2503,168 @@ def test_sum_X(tmpdir, dist_mem):
+    if not dist_mem:
+        output = (
+            "    SUBROUTINE invoke_0(f2, a, f1)\n"
+            "      REAL(KIND=r_def), intent(in) :: a\n"
```
Done.

TeranIvy commented on src/psyclone/tests/domain/lfric/lfric_builtins_test.py:
```diff
@@ -2058,15 +2503,168 @@ def test_sum_X(tmpdir, dist_mem):
+
+    if not dist_mem:
+        output = (
+            "    SUBROUTINE invoke_0(f2, f1)\n"
```
As this built-in works with `real` and `integer` fields, I'd like to leave the declarations and all to make visible what are the correct argument types, argument proxy types and `use` statements.

arporter replied:
OK.

TeranIvy commented on src/psyclone/tests/domain/lfric/lfric_integer_builtins_test.py:
```diff
@@ -187,6 +187,141 @@ def test_int_inc_X_plus_Y(tmpdir, monkeypatch, annexed, dist_mem):
+
+    if not dist_mem:
+        output = (
+            "    SUBROUTINE invoke_0(f2, a, f1)\n"
```
Yep, done.

TeranIvy commented on src/psyclone/tests/domain/lfric/lfric_integer_builtins_test.py:
```diff
@@ -702,3 +837,156 @@ def test_int_setval_X(tmpdir, monkeypatch, annexed, dist_mem):
+
+    if not dist_mem:
+        output = (
+            "    SUBROUTINE invoke_0(f2, a, f1)\n"
```
Done.

TeranIvy commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -1807,7 +1807,7 @@ def test_builtin_loop_fuse_do(tmpdir, monkeypatch, annexed, dist_mem):
 
```
Fixed re-imports and wrong import order.

arporter replied:
Thanks.

TeranIvy commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -239,6 +245,17 @@ def _validate(self):
+            data_types_str = [str(x) for x in sorted(data_types)]
+            raise ParseError(
+                "In the LFRic API only the data type conversion built-ins "
+                "{0} are allowed to have different data types of their "
```
This reads better, thanks. Changed.

TeranIvy commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -862,6 +992,64 @@ def gen_code(self, parent):
+
+class LFRicADividebyXKern(LFRicBuiltIn):
+    ''' DoF-wise division of a scalar value `a` by the elements of
+    a real-valued field, `X`, and storing the result in another,
```
Done.

TeranIvy commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -862,6 +992,64 @@ def gen_code(self, parent):
+
+class LFRicIncADividebyXKern(LFRicBuiltIn):
+    ''' DoF-wise division of a scalar value `a` by the elements of
+    a real-valued field, `X`, and storing the result in the same
```
Done.

TeranIvy commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -862,6 +992,64 @@ def gen_code(self, parent):
+
+        '''
+        # We divide the real scalar argument by each element of f1 and
+        # return the result in f1 (real-valued fields).
```
Done.

TeranIvy commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -1049,10 +1237,79 @@ def gen_code(self, parent):
+
+
+class LFRicSignXKern(LFRicBuiltIn):
+    ''' Returns the sign of a real-valued field elements using the
```
You're probably right, it can be googled. Removed the explanation.

TeranIvy commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -1049,10 +1237,79 @@ def gen_code(self, parent):
+
+
+class LFRicReal2IntXKern(LFRicBuiltIn):
+    ''' Converts a real-valued field elements to an integer-valued
```
Done, thanks.

TeranIvy commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -1200,6 +1481,60 @@ def __str__(self):
+class LFRicIntSignXKern(LFRicSignXKern):
+    ''' Returns the sign of an integer-valued field elements using the
+    Fortran intrinsic `sign` function, `Y = sign(a, X)`, where `a` is
+    an integer scalar and `Y` and `X` are integer-valued fields. The
```
Removed the explanation.

TeranIvy commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -1200,6 +1481,60 @@ def __str__(self):
+
+
+class LFRicInt2RealXKern(LFRicBuiltIn):
+    ''' Converts an integer-valued field elements to a real-valued
```
Done.

TeranIvy commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -1241,13 +1584,19 @@ def __str__(self):
     # Sum values of a real field
-    "sum_X": LFRicSumXKern}
+    "sum_X": LFRicSumXKern,
+    # Sign of real field elements
```
Done.

TeranIvy commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -1260,7 +1609,11 @@ def __str__(self):
     "int_setval_c": LFRicIntSetvalCKern,
-    "int_setval_X": LFRicIntSetvalXKern}
+    "int_setval_X": LFRicIntSetvalXKern,
+    # Sign of integer field elements
```
Done.

TeranIvy commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -1049,10 +1237,79 @@ def gen_code(self, parent):
+        :type parent: :py:class:`psyclone.f2pygen.BaseGen`
+
+        '''
+        # Return the sign of all the elements of a real-valued field using
```
Yes, that reads better. Changed.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2184,6 +2189,11 @@ scheme presented below. Any new Built-in needs to comply with these rules.
       operations on the ``integer``-valued fields.
 
+7) As in the case of Built-in field argument rules, the names of
+   the field data type conversion Built-ins, ``real2int_X`` and
```
Done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2621,6 +2726,56 @@ where:
+Sign of elements
+################
+
+A Built-in which returns the sign of a ``real``-valued field elements
```
Done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2621,6 +2726,56 @@ where:
+
+**sign_X** (*field2*, *rscalar*, *field1*)
+
+Returns the sign of a ``real``-valued field elements using the Fortran
```
Done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2621,6 +2726,56 @@ where:
+Returns the sign of a ``real``-valued field elements using the Fortran
+intrinsic ``sign`` function as ``Y = sign(a, X)``, where ``a`` is a
+``real`` scalar and ``Y`` and ``X`` are ``real``-valued fields.
+The results are ``a`` for ``a > 0``, ``0`` for ``a = 0`` and ``-a``
```
I mixed things up in my test program. It should actually be: "results are ``a`` for ``X >= 0`` and ``-a`` for ``X < 0``" so I corrected it.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2621,6 +2726,56 @@ where:
+Conversion of ``real`` to ``integer`` field elements
+####################################################
+
+A Built-in which takes a ``real`` field elements and converts them to
```
Yep, done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2621,6 +2726,56 @@ where:
+####################################################
+
+A Built-in which takes a ``real`` field elements and converts them to
+an ``integer`` field elements is denoted with the keyword **real2int**.
```
Done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2621,6 +2726,56 @@ where:
+
+**real2int_X** (*ifield2*, *field1*)
+
+Converts a ``real``-valued field elements to an ``integer``-valued
```
Done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2802,6 +2987,56 @@ where:
+################
+
+A Built-in which returns the sign of an ``integer``-valued field
+elements is denoted with the keyword **sign** and the prefix **int**.
```
Done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2802,6 +2987,56 @@ where:
+Sign of elements
+################
+
+A Built-in which returns the sign of an ``integer``-valued field
```
Re-written as for the `real` equivalent above.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2802,6 +2987,56 @@ where:
+Returns the sign of an ``integer``-valued field elements using the
+Fortran intrinsic ``sign`` function as ``Y = sign(a, X)``, where ``a``
+is an ``integer`` scalar and ``Y`` and ``X`` are ``integer``-valued
+fields. The results are ``a`` for ``a > 0``, ``0`` for ``a = 0`` and
```
Yep, done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2802,6 +2987,56 @@ where:
+Conversion of ``integer`` to ``real`` field elements
+####################################################
+
+A Built-in which takes an ``integer`` field elements and converts them
```
Yep, done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2802,6 +2987,56 @@ where:
+####################################################
+
+A Built-in which takes an ``integer`` field elements and converts them
+to a ``real`` field elements is denoted with the keyword **int2real**.
```
Done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2802,6 +2987,56 @@ where:
+
+**int2real_X** (*field2*, *ifield1*)
+
+Converts an ``integer``-valued field elements to a ``real``-valued
```
Done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2036,7 +2036,12 @@ following rules:
    ``integer`` scalar arguments;
 
 7) Built-ins that update ``integer``-valued fields can only read from
```
Good point, done.

TeranIvy commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -2184,6 +2189,11 @@ scheme presented below. Any new Built-in needs to comply with these rules.
+7) As in the case of Built-in field argument rules, the names of
+   the field data type conversion Built-ins, ``real2int_X`` and
+   ``int2real_X``, are the only exceptions to the rules for
+   Built-ins names in Fortran above.
```
Thanks for the suggestion, done.

TeranIvy commented on src/psyclone/tests/domain/lfric/lfric_builtins_test.py:
```diff
@@ -2058,15 +2503,168 @@ def test_sum_X(tmpdir, dist_mem):
+# ------------- Converting real to integer field elements ------------------- #
+
+
+def test_real2int_X(tmpdir, monkeypatch, annexed, dist_mem):
```
Good point! As discussed at the PSyclone Telco, I simply called the conversion built-ins as the respective intrinsics they are using, i.e. `int_X` for the `real` to `integer` conversion and `real_X` for `real` to `integer`. This is now also updated in the documentation.

TeranIvy commented:
Addressed the review comments and brought the branch up to date with the master. Ready for another review from @arporter.

arporter reviewed (commented):
All requested changes have been made.
All tests and examples OK with compilation.
Docs all build OK although I note we now have a warning in transformations.rst, however,  that's not the fault of this PR.
I would approve but will await confirmation from @TeranIvy that LFRic works OK with these new built-ins.

TeranIvy commented:
@arporter, thanks for making me do the check as there was a reduction access for the scalar argument in code generation for `sign_X` that we both missed. This is now corrected.

Besides `sign_X`, I modified some more algorithms to test the conversion Built-ins, `int_X` and `real_X`. I also tested `a_divideby_X` in one of the key algorithms that currently needs to create a field populated with `1.0_r_def` to calculate an inverse of another field. In addition, I also tested `int_inc_X_plus_Y`, `int_setval_c` and `int_X_times_Y`.

The test suite is now fine. Back to review.

arporter commented:
Test suite, examples and docs still all OK. Will proceed to merge once CI is complete.
